### PR TITLE
Fix Failover SSL

### LINKED ISSUE #798: Failover does not respect individual failover cluster SSL settings [API-1412]
#### Expected behaviour

Each SSL setting in the individual failover cluster items (`sslConfig1` and `sslConfig2`) should be respected when trying to connect to the cluster:
```
config := hazelcast.Config{}
c1 := cluster.Config{}
c1.Network.SSL = sslConfig1
c2 := cluster.Config{}
c1.Network.SSL = sslConfig2
config.Failover.Enabled = true
config.Failover.SetConfigs(c1, c2)
client, err := hazelcast.StartNewClientWithConfig(context.Background(), config)
```

#### Actual behaviour

Only the SSL setting in the root `Cluster` configuration is used.

### PR DESCRIPTION
**Pull request by yuce**
* Fixes SSL configuration in failover cluster configuration not honored. Fixes #798
* Makes authentication errors retryable.

### CODE REVIEW

File: internal/cluster/connection_manager.go [hunk 1]
```diff
@@ -463,11 +462,13 @@ func (m *ConnectionManager) connectCluster(ctx context.Context, cluster *Candida
 	return initialAddr, nil
 }
 
-func (m *ConnectionManager) ensureConnection(ctx context.Context, addr pubcluster.Address) (*Connection, error) {
+func (m *ConnectionManager) ensureConnection(ctx context.Context, addr pubcluster.Address, networkCfg *pubcluster.NetworkConfig) (*Connection, error) {
 	conn := m.createDefaultConnection()
-	if err := conn.start(m.clusterConfig, addr); err != nil {
+	err := conn.start(networkCfg, addr)
+	if err != nil {
```
Comment (utku-caglayan) [inline internal/cluster/connection_manager.go:468]: Can we merge this into one line with the statement above?

File: internal/cluster/connection_manager.go [hunk 2]
```diff
@@ -564,9 +564,9 @@ func (m *ConnectionManager) processAuthenticationResult(conn *Connection, result
 		})
 		return conn, nil
 	case credentialsFailed:
-		return nil, cb.WrapNonRetryableError(fmt.Errorf("invalid credentials: %w", hzerrors.ErrAuthentication))
+		return nil, fmt.Errorf("invalid credentials: %w", hzerrors.ErrAuthentication)
```
Comment (utku-caglayan) [inline internal/cluster/connection_manager.go:567]: Why do we change this to retryable?

File: internal/cluster/connection_manager_test.go [hunk 3]
```diff
@@ -95,18 +111,17 @@ func tryConnectAddressTest(checkedAddresses []CheckedAddressHelper, inputAddress
...
-	m.clusterConfig.Network.PortRange = portRange
-	resultAddr, err := m.tryConnectAddress(context.TODO(), pubcluster.NewAddress(host, int32(port)),
-		func(ctx context.Context, m *ConnectionManager, currAddr pubcluster.Address) (pubcluster.Address, error) {
-			connMemberCounter++
-			for _, checkedAddr := range checkedAddresses {
-				if currAddr == checkedAddr.address {
-					return checkedAddr.address, checkedAddr.err
-				}
+	nc := &pubcluster.NetworkConfig{PortRange: portRange}
+	mf := func(ctx context.Context, m *ConnectionManager, currAddr pubcluster.Address, networkCfg *pubcluster.NetworkConfig) (pubcluster.Address, error) {
```
Comment (utku-caglayan) [inline internal/cluster/connection_manager_test.go:115]:
```suggestion
	mf := func(_ context.Context, _ *ConnectionManager, currAddr pubcluster.Address, _ *pubcluster.NetworkConfig) (pubcluster.Address, error) {
```

File: internal/cluster/connection_manager.go (hunk 2 above)
Comment (utku-caglayan) [reply internal/cluster/connection_manager.go:567]: Ignore this comment please

File: .gitignore [hunk 4]
```diff
@@ -31,4 +31,7 @@ vet.txt
 # Test artifacts
 test.pid
 coverage.html
-coverage.out
\ No newline at end of file
+coverage.out
+
+# Others
+cmd/
```
[the remark this answers is not in the document]
Comment (yuce) [reply .gitignore:37]: That's a personal folder I keep small apps to test features.

File: cluster/failover_config.go [hunk 5]
```diff
@@ -112,6 +112,15 @@ func (c *FailoverConfig) Validate(root Config) error {
 				i, hzerrors.ErrIllegalArgument)
 		}
 	}
+	// root SSL configuration is not respected if failover is enabled.
```
[the remark this answers is not in the document]
Comment (yuce) [reply cluster/failover_config.go:115]: Enabling SSL configuration in the failover doesn't matter. If failover is enabled, SSL configurations in the failover cluster is used even if SSL configuraiton was not enabled in those cluster options.

File: cluster/failover_config.go [hunk 6]
```diff
@@ -112,6 +112,15 @@ func (c *FailoverConfig) Validate(root Config) error {
...
+	// check that root SSL configuration is not enabled if one of the failover configs is disabled.
```
[the remark this answers is not in the document]
Comment (yuce) [reply cluster/failover_config.go:116]: The changed text is not correct.

File: cluster/failover_config.go [hunk 7]
```diff
@@ -112,6 +112,15 @@ func (c *FailoverConfig) Validate(root Config) error {
...
+	if root.Network.SSL.Enabled {
+		for _, c := range c.Configs {
```
[the remark this answers is not in the document]
Comment (yuce) [reply cluster/failover_config.go]: Updated at: https://github.com/hazelcast/hazelcast-go-client/pull/797/commits/f1f54ae444f68d631c89d6634699f10ff0898a1f

File: cluster/failover_config.go [hunk 8]
```diff
@@ -112,6 +112,15 @@ func (c *FailoverConfig) Validate(root Config) error {
...
+			if !c.Network.SSL.Enabled {
+				return fmt.Errorf("root SSL configuration cannot be enabled if failover is enabled, use failover configurations instead: %w", hzerrors.ErrInvalidConfiguration)
```
[the remark this answers is not in the document]
Comment (yuce) [reply cluster/failover_config.go]: Updated at: https://github.com/hazelcast/hazelcast-go-client/pull/797/commits/f1f54ae444f68d631c89d6634699f10ff0898a1f

File: internal/cluster/connection_manager.go (hunk 1 above)
Comment (yuce) [reply internal/cluster/connection_manager.go]: Updated at: https://github.com/hazelcast/hazelcast-go-client/pull/797/commits/f1f54ae444f68d631c89d6634699f10ff0898a1f

File: internal/cluster/connection_manager_test.go (hunk 3 above)
Comment (yuce) [reply internal/cluster/connection_manager_test.go]: Updated at:" https://github.com/hazelcast/hazelcast-go-client/pull/797/commits/f1f54ae444f68d631c89d6634699f10ff0898a1f

File: cluster/failover_config.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (utku-caglayan) [reply cluster/failover_config.go:116]: What does "if one of the failover config is disabled" stands for? Does it specify "one of the failover config that has SSL disabled"?

[the remark this answers is not in the document]
Comment (yuce) [reply cluster/failover_config.go]: That's right, updated at: https://github.com/hazelcast/hazelcast-go-client/pull/797/commits/83e5665baeb9d4b8e4a5d1fda091b5e64fae4eeb

**Review by nevzatseferoglu — APPROVED** [review]
Thanks, looks good to me!

**Comment by yuce** [comment]
Report coverage test failed with the following log:
```
2022-06-24T11:39:39.1796986Z === RUN   TestClusterShutdownThenCheckOperationsNotHanging
2022-06-24T11:39:39.1797463Z === RUN   TestClusterShutdownThenCheckOperationsNotHanging/Smart_Client
2022-06-24T11:39:39.1797972Z 2022/06/24 11:39:15 INFO : trying to connect to cluster: invocation-after-shutdown-2-true
2022-06-24T11:39:39.1798407Z 2022/06/24 11:39:15 ERROR: cannot connect to localhost:44701: dial tcp [::1]:44701: connect: connection refused: dial tcp [::1]:44701: connect: connection refused
2022-06-24T11:39:39.1798929Z 2022/06/24 11:39:16 ERROR: cannot connect to localhost:44701: dial tcp [::1]:44701: connect: connection refused: dial tcp [::1]:44701: connect: connection refused
2022-06-24T11:39:39.1799398Z 2022/06/24 11:39:17 ERROR: cannot connect to localhost:44701: dial tcp [::1]:44701: connect: connection refused: dial tcp [::1]:44701: connect: connection refused
2022-06-24T11:39:39.1799870Z 2022/06/24 11:39:18 ERROR: cannot connect to localhost:44701: dial tcp [::1]:44701: connect: connection refused: dial tcp [::1]:44701: connect: connection refused
2022-06-24T11:39:39.1800338Z 2022/06/24 11:39:19 ERROR: cannot connect to localhost:44701: dial tcp [::1]:44701: connect: connection refused: dial tcp [::1]:44701: connect: connection refused
2022-06-24T11:39:39.1800797Z 2022/06/24 11:39:20 ERROR: cannot connect to localhost:44701: dial tcp [::1]:44701: connect: connection refused: dial tcp [::1]:44701: connect: connection refused
2022-06-24T11:39:39.1801263Z 2022/06/24 11:39:22 ERROR: cannot connect to localhost:44701: dial tcp [::1]:44701: connect: connection refused: dial tcp [::1]:44701: connect: connection refused
2022-06-24T11:39:39.1801713Z 2022/06/24 11:39:23 ERROR: cannot connect to localhost:44701: dial tcp [::1]:44701: connect: connection refused: dial tcp [::1]:44701: connect: connection refused
2022-06-24T11:39:39.1802170Z 2022/06/24 11:39:24 ERROR: cannot connect to localhost:44701: dial tcp [::1]:44701: connect: connection refused: dial tcp [::1]:44701: connect: connection refused
2022-06-24T11:39:39.1802639Z 2022/06/24 11:39:26 ERROR: cannot connect to localhost:44701: dial tcp [::1]:44701: connect: connection refused: dial tcp [::1]:44701: connect: connection refused
2022-06-24T11:39:39.1803107Z 2022/06/24 11:39:27 ERROR: cannot connect to localhost:44701: dial tcp [::1]:44701: connect: connection refused: dial tcp [::1]:44701: connect: connection refused
2022-06-24T11:39:39.1803564Z 2022/06/24 11:39:29 ERROR: cannot connect to localhost:44701: dial tcp [::1]:44701: connect: connection refused: dial tcp [::1]:44701: connect: connection refused
2022-06-24T11:39:39.1804031Z 2022/06/24 11:39:31 ERROR: cannot connect to localhost:44701: dial tcp [::1]:44701: connect: connection refused: dial tcp [::1]:44701: connect: connection refused
2022-06-24T11:39:39.1804489Z 2022/06/24 11:39:33 ERROR: cannot connect to localhost:44701: dial tcp [::1]:44701: connect: connection refused: dial tcp [::1]:44701: connect: connection refused
2022-06-24T11:39:39.1804928Z 2022/06/24 11:39:34 ERROR: cannot connect to localhost:44701: dial tcp [::1]:44701: connect: connection refused: dial tcp [::1]:44701: connect: connection refused
2022-06-24T11:39:39.1805434Z --- FAIL: TestClusterShutdownThenCheckOperationsNotHanging (34.73s)
2022-06-24T11:39:39.1806045Z     --- FAIL: TestClusterShutdownThenCheckOperationsNotHanging/Smart_Client (34.73s)
2022-06-24T11:39:39.1806451Z panic: cannot connect to any cluster: deadline exceeded [recovered]
2022-06-24T11:39:39.1806783Z 	panic: cannot connect to any cluster: deadline exceeded
2022-06-24T11:39:39.1806945Z 
2022-06-24T11:39:39.1807028Z goroutine 2650 [running]:
2022-06-24T11:39:39.1807287Z testing.tRunner.func1.1(0x8fc980, 0xc0001f3ce0)
2022-06-24T11:39:39.1807646Z 	/opt/hostedtoolcache/go/1.15.15/x64/src/testing/testing.go:1072 +0x30d
2022-06-24T11:39:39.1807931Z testing.tRunner.func1(0xc0003fa180)
2022-06-24T11:39:39.1808248Z 	/opt/hostedtoolcache/go/1.15.15/x64/src/testing/testing.go:1075 +0x41a
2022-06-24T11:39:39.1808500Z panic(0x8fc980, 0xc0001f3ce0)
2022-06-24T11:39:39.1808802Z 	/opt/hostedtoolcache/go/1.15.15/x64/src/runtime/panic.go:969 +0x1b9
2022-06-24T11:39:39.1809398Z github.com/hazelcast/hazelcast-go-client/internal/it.MustClient(...)
2022-06-24T11:39:39.1809916Z 	/home/runner/work/hazelcast-go-client/hazelcast-go-client/internal/it/util.go:197
2022-06-24T11:39:39.1810562Z github.com/hazelcast/hazelcast-go-client_test.TestClusterShutdownThenCheckOperationsNotHanging.func1(0xc0003fa180, 0x9321b11601)
2022-06-24T11:39:39.1811275Z 	/home/runner/work/hazelcast-go-client/hazelcast-go-client/client_it_test.go:780 +0x653
2022-06-24T11:39:39.1811754Z github.com/hazelcast/hazelcast-go-client_test.clientTester.func1(0xc0003fa180)
2022-06-24T11:39:39.1812263Z 	/home/runner/work/hazelcast-go-client/hazelcast-go-client/client_it_test.go:822 +0x34
2022-06-24T11:39:39.1812581Z testing.tRunner(0xc0003fa180, 0xc0005c21b0)
2022-06-24T11:39:39.1812900Z 	/opt/hostedtoolcache/go/1.15.15/x64/src/testing/testing.go:1123 +0xef
2022-06-24T11:39:39.1813155Z created by testing.(*T).Run
2022-06-24T11:39:39.1813458Z 	/opt/hostedtoolcache/go/1.15.15/x64/src/testing/testing.go:1168 +0x2b3
2022-06-24T11:39:39.1813858Z FAIL	github.com/hazelcast/hazelcast-go-client	435.298s
```

Line 780 corresponds to:
```
client := it.MustClient(hz.StartNewClientWithConfig(ctx, config))
```
So the client couldn't connect. I believe this is an RC-related problem.

**Review by utku-caglayan — APPROVED** [review]

**Review by nevzatseferoglu — APPROVED** [review]

**Pull request merged**